### ISSUE TITLE
Can I use tiny-cnn for classifying human and car?

### ISSUE THREAD

**Issue by younglittlefat** [tiny-dnn/tiny-dnn#18]
Dear nyanp:
  Tiny-cnn is a CNN of smart and excellent which can be used in Windows. I've been studying to it for several days, and I got some problems recently.

  I try to use tiny-cnn for classifying human and car. I label images of human to 0 and car to 1. These images are convert to gray-scale and resize to 32x32. Default setting has been used but the output dimension of the last fully connected layer has been changed to 2.

  All images are took from surveillant video, but images of human are from one camera and cars are from another. The subject occupies the main space in every image.

  Now something strange happens. All the outputs of testing set are 1, i.e. car. I have no idea wether problem comes from data set or settings of tiny-cnn. 

  If more detail is needed, I'll reply as far as possible. Thanks very much!

 Best regards,
 younglittlefat

**Comment by nyanp** [comment]
Dear younglittlefat,

- Is your training data randomly shuffled? Result of training neural networks depends on the data ordering.
- Is your data highly unbalanced? For example, if images of car are 99% of data, CNN may assign everything to car.

see also:
[http://ufldl.stanford.edu/tutorial/supervised/OptimizationStochasticGradientDescent/](http://ufldl.stanford.edu/tutorial/supervised/OptimizationStochasticGradientDescent/)
[http://www.quora.com/How-can-you-train-convolutional-neural-networks-on-highly-unbalanced-datasets](http://www.quora.com/How-can-you-train-convolutional-neural-networks-on-highly-unbalanced-datasets)

**Comment by GangFang** [comment]
Dear nyanp,
First thanks for your great work, it really facilitate my research (:

And I come across the SAME problem as younglittlefat's.

Originally I am going to classify traffic signs into 13 classes. I first built a small training set (for testing, totally 1500 images) but the CNN classify them all into one class.
Considering the training set may be too small to test the model, I then randomly selected 2 classes from those 13 traffic signs classes with totally 50,000 images (nearly 25,000 vs 25,000) and convert them into .idx format WITH MNISTEN. According to the document, it will handle shuffling and resizing automatically.  

So do you have any idea on this problem? Since the data has been randomly shuffled by MNISTEN and it is not unbalanced, I am considering what is the problem here.

Any comment will be greatly appreciated. THX!!

**Comment by GangFang** [comment]
Yup, just forgot to mention that I use MNISTEN to create a test set with 15,000 images from those 50,000 images (so the training set includes roughly 17,500 class 1 images and 17,500 class 2 images). 

And the result is always 8,486/15,000.

**Comment by nyanp** [comment]
@GangFang 
Good parameters for classifying depends on your task, so let's try another parameters. For example:

- set smaller learning rates (alpha)
- change optimizers, i.e. adagrad
- change mini-batch size

```cpp
network<mse, adagrad> nn;

// change learning rate
nn.optimizer().alpha *= 0.1;
```

**Comment by edgarriba** [comment]
@gangfang Did you solve this?

**Comment by Randl** [comment]
@younglittlefat Have you succeeded?

**Comment by Randl** [comment]
No answer for a week, closing
